### DISCUSSION TITLE
"The passed message type is invalid" with Jetson Nano 4Gb + ROS 2 Galactic container

### DISCUSSION THREAD

**Discussion by ParadoxRobotics** [iRobotEducation/create3_docs#160] [Questions]
Hi, 

I currently building a Turtlebot like robot with the Create3 ^ ^
I use a Jetson nano 4Gb (rev A) developper kit as the main processing unit connected in USB to the create3 and powered by an external battery.
I followed the instruction to set up my system : https://iroboteducation.github.io/create3_docs/setup/jetson/
But unfortunately, for every action (like dock/undock, audio, ...) I get : 
```
root@MARVIN01:/# ros2 topic pub /cmd_audio irobot_create_msgs/msg/AudioNoteVector "{append: false, notes: [{frequency: 392, max_runtime: {sec: 0,nanosec: 177500000}}, {frequency: 523, max_runtime: {sec: 0,nanosec: 355000000}}, {frequency: 587, max_runtime: {sec: 0,nanosec: 177500000}}, {frequency: 784, max_runtime: {sec: 0,nanosec: 533000000}}]}" -1
The passed message type is invalid

```
It seems the irobot_create_msgs format is the issue, But I don't know how to solve this...
How can I build the https://github.com/iRobotEducation/irobot_create_msgs package in the container with ROS 2 ?
Is it specified in the documentation ?

Best regards

**Comment by alsora** [comment]
Yes, as you said, the problem is with the `irobot_create_msgs` package that is missing from your environment.

That's a regular ROS package, you can either build it from sources or install it from debian packages.

To install from Debian, just run this (you can also add these lines to your Dockerfile to avoid having to repeat them everytime)
```
sudo apt-get update
sudo apt-get install ros-galactic-irobot-create-msgs
```

To build from sources follow the instructions here https://docs.ros.org/en/foxy/Tutorials/Beginner-Client-Libraries/Creating-Your-First-ROS2-Package.html#build-a-package

**Reply by ParadoxRobotics** [reply]
Thanks for your answer.
I think this should be in the documentation !

**Reply by alsora** [reply]
Thanks for reporting it.
Indeed this was missing from the Jetson nano instructions.

I fixed it and the updated version should show up on the webiste shortly

**Reply by ParadoxRobotics** [reply]
The installation from the Debian package fail inside the container : 

```
root@MARVINN:/# sudo apt update && sudo apt install -y ros-galactic-ros-base python3-colcon-common-extensions python3-rosdep ros-galactic-rmw-fastrtps-cpp ros-galactic-rmw-cyclonedds-cpp ros-galactic-irobot-create-msgs
Get:1 http://ports.ubuntu.com/ubuntu-ports bionic InRelease [242 kB]
Get:2 http://ports.ubuntu.com/ubuntu-ports bionic-updates InRelease [88.7 kB]  
Get:3 http://ports.ubuntu.com/ubuntu-ports bionic-backports InRelease [74.6 kB]
Get:4 http://ports.ubuntu.com/ubuntu-ports bionic-security InRelease [88.7 kB] 
Get:5 https://apt.kitware.com/ubuntu bionic InRelease [11.0 kB]                
Get:6 http://packages.ros.org/ros2/ubuntu bionic InRelease [4,680 B]     
Get:7 http://ports.ubuntu.com/ubuntu-ports bionic/universe arm64 Packages [11.0 MB]
Get:8 http://ports.ubuntu.com/ubuntu-ports bionic/main arm64 Packages [1,285 kB]
Get:9 http://ports.ubuntu.com/ubuntu-ports bionic/multiverse arm64 Packages [153 kB]
Get:10 http://ports.ubuntu.com/ubuntu-ports bionic/restricted arm64 Packages [572 B]
Get:11 http://ports.ubuntu.com/ubuntu-ports bionic-updates/universe arm64 Packages [2,063 kB]
Get:12 http://ports.ubuntu.com/ubuntu-ports bionic-updates/multiverse arm64 Packages [5,351 B]
Get:13 http://ports.ubuntu.com/ubuntu-ports bionic-updates/restricted arm64 Packages [5,451 B]
Get:14 http://ports.ubuntu.com/ubuntu-ports bionic-updates/main arm64 Packages [2,018 kB]
Get:15 http://ports.ubuntu.com/ubuntu-ports bionic-backports/main arm64 Packages [12.2 kB]
Get:16 http://ports.ubuntu.com/ubuntu-ports bionic-backports/universe arm64 Packages [12.5 kB]
Get:17 http://ports.ubuntu.com/ubuntu-ports bionic-security/multiverse arm64 Packages [2,761 B]
Get:18 http://ports.ubuntu.com/ubuntu-ports bionic-security/universe arm64 Packages [1,366 kB]
Get:19 http://ports.ubuntu.com/ubuntu-ports bionic-security/restricted arm64 Packages [4,778 B]
Get:20 http://ports.ubuntu.com/ubuntu-ports bionic-security/main arm64 Packages [1,639 kB]
Get:21 https://apt.kitware.com/ubuntu bionic/main arm64 Packages [30.1 kB]     
Get:22 http://packages.ros.org/ros2/ubuntu bionic/main arm64 Packages [660 kB] 
Fetched 20.8 MB in 4s (4,857 kB/s)                        
Reading package lists... Done
Building dependency tree       
Reading state information... Done
181 packages can be upgraded. Run 'apt list --upgradable' to see them.
Reading package lists... Done
Building dependency tree       
Reading state information... Done
E: Unable to locate package ros-galactic-ros-base
E: Unable to locate package ros-galactic-rmw-fastrtps-cpp
E: Unable to locate package ros-galactic-rmw-cyclonedds-cpp
E: Unable to locate package ros-galactic-irobot-create-msgs
```

Should I build from source ?

**Reply by alsora** [reply]
Yes, you can build from source,  but it should not be necessary. That error indicates that Debian sources are not correctly set in the container.

Which docker image are you using? Can you give me the full name?

**Reply by ParadoxRobotics** [reply]
It's the one provided in the documentation : ros:galactic-ros-base-l4t-r32.6.1
I will try with a more recent one !

**Reply by ParadoxRobotics** [reply]
I use a more recent container -> ros:galactic-pytorch-l4t-r35.1.0 (with Pytorch because I need it for an experiment)
I can Install all the dependencies, but unfortunately I still got : 
```
root@MARVINN:/# ros2 topic pub /cmd_audio irobot_create_msgs/msg/AudioNoteVector "{append: false, notes: [{frequency: 392, max_runtime: {sec: 0,nanosec: 177500000}}, {frequency: 523, max_runtime: {sec: 0,nanosec: 355000000}}, {frequency: 587, max_runtime: {sec: 0,nanosec: 177500000}}, {frequency: 784, max_runtime: {sec: 0,nanosec: 533000000}}]}" -1
The passed message type is invalid

```

**Reply by alsora** [reply]
Have you sourced the ROS 2 setup scripts after adding new packages?
Assuming that you installed from apt-get, this should be
```
source /opt/ros/galactic/setup.sh
```

Unfortunately I don't have a Jetson nano at hand, but I just checked this in ROS 2 docker container and, after installing the `ros-galactic-irobot-create-msgs` package and sourcing the setup script, I'm able to use all Create 3 custom ROS 2 messages.

**Reply by ParadoxRobotics** [reply]
Hi, 
thanks for yout answer.
I create my own image for the Jetson nano that support Ubuntu 20.04 and ROS2 : It work on the robot with no problem.
I will try your idea with my other jetson in the next day.

**Reply by bwmcin** [reply]
ParadoxRobotics, did you every try the idea with your other Jetson and did it work?  Also, for the Ubuntu 20.04 which install did you do?  I am just about ready to start once I get my Create 3 but want to prep ahead.  Thank you.
